### DISCUSSION TITLE
[Proposal] Language "Dialects"

### DISCUSSION THREAD

**Discussion by En3Tho** [dotnet/csharplang#5012] [Language Ideas] [answered]
I'ts just a thought on how can we enforce a "modern" C# features in new projects. So basically "dialect" is a set of up-to-date modern features which are recommended to use in new projects. Dialects do not add any new features, they only constrain them to the recommended ones. Or you can say that it is "dialect-local" deprecation of features. So if you just want to bump a legacy project to newest C# version then it's no problem. But for new project people could use a new dialect which will force them to drop old unwanted/unpopular features in their new code.

The most simple one is null checking: is/is not instead of ==
Top level namespaces instead of ones with nesting
Simplified parameter null checking instead of explicit ones
etc

It's like a light deprecation policy. I understand that C# team can't really deprecate features, but every new C# version could have a dialect directly associated with it.

Again a new project example: C# 10 comes out and I create a new project. I set dialect version to 10 too to deprecate old features from old C#'s. Year after C# 11 comes out and I set language version to 11 and nothing breaks because dialect is still set to 10. In my new project I can use both lang and dialect versions set to 11 and carry on with only latest and recommended features. If I choose to set dialect 11 in that old project then I get build warnings (which can be set to errors) which I have to fix.

Ideally, as team knows which features are considered old and deprecated with each version, a refactoring fix could be provided: migrate from dialect xxx to yyy. I don't know whether it is realizable or not.

Dialects could come with a new C# version or with every LTS .Net Release.

Pros I see: a mild way of deprecating things and setting code preferences straight for everyone. Reducing language bloat.

Cons: a potential blow up of dialects. People are not forced to do it but it could be a problem? Like some people could not like the new dialect (they REALLY loved a feature from the old one) so they decide to stay on it. Also people will have to track which was deprecated in which dialect so there should be a help page which could thoroughly describe all deprecation changes from dialect xxx to dialect yyy.

**Comment by yaakov-h** [comment]
If you want to restrict code to a subset of legal code, this is the exact purpose of Roslyn Analyzers.

**Reply by En3Tho** [reply]
There is no official analyzer helping with features really. There is no feature based apis, only pattern based. So you have to describe every feature with a set of patterns and write the analyzer from ground up, not to mention keep it up to date. If such analyzer existed I believe it would be a job of Roslyn/C# team to keep it up to date. But if C# feels like new features do not deprecate old ones then I've got my answer.

**Reply by Joe4evr** [reply]
> There is no official analyzer helping with features really.

But there are. Open up an older project and set it to target .NET 5 with C# 9, and you'll get plenty of feature suggestions. "Use pattern matching", "Use target-typed new", "Use range operator", etc. There's only a few like that, but they ship in the box, and if you think there's some language feature that should be preferred over a different way of doing the same thing, you can write an analyzer for that yourself.

**Reply by En3Tho** [reply]
You mean a refactoring analyzer/code fixer. Obviously they exist. What I mean is more of constraining analyzer. Like something you could enforce in a company/team. Maybe I'm wrong but I haven't seen an official one like this.

**Reply by En3Tho** [reply]
Just wanted to write a big reply and one small occasional movement and GitHub mobile sent it to hell. -_-.

Well, It was just an idea. If it turned out to be bad or not to people liking then it's ok. Thank you for your response.

**Reply by Joe4evr** [reply]
> What I mean is more of constraining analyzer. Like something you could enforce in a company/team. Maybe I'm wrong but I haven't seen an official one like this.

Well yes, but that's because the C# language design team has always maintained a position of "if it compiles today, it should compile tomorrow" and that applies to all the code written over the past 20 years, syntactic sugar be damned.

If you want to enforce "older" features/patterns may not be used for your entire team/company, then you can easily write your own analyzers for that, because that's exactly what that entire infrastructure is there for, and you've explicitly opted in to receiving your new errors any time they go off.

**Reply by CyrusNajmabadi** [reply]
> Well, It was just an idea. If it turned out to be bad or not to people liking then it's ok. Thank you for your response.

It's not a bad idea at all. I think the general feedback here is simply that there's already a good solution for this. So there's no need for some other solution to be made that solves it in another fashion. Thanks!

**Comment by HaloFour** [comment, marked as answer]
The language team has been pretty opposed to the idea of adding dialects to the language.  Even with NRTs the team was very careful that the dialect had no impact on the behavior of the compiled code, at most it might generate a few additional warnings.

The team would also disagree that newer features deprecate older ones and that there would be _very_ few language features that they consider no longer useful.

**Reply by En3Tho** [reply]
I see.

**Reply by En3Tho** [reply]
Can you please elaborate on "very few"? I'm really curious to know which features lang team doesn't consider well, maybe "really needed".

**Reply by HaloFour** [reply]
"Anonymous delegates" are probably the highest on that list.  They're not completely useless but most cases are better served with lambdas.  I'd almost wager extension methods as well, not because there is a better option but because the existing syntax doesn't leave much room to evolve the feature and "extension everything" has been exploring options which would subsume the existing functionality into a different syntax.

I'm sure some language team members would chime in that even those cases aren't "deprecated".

**Reply by En3Tho** [reply]
Thanks
